### PR TITLE
Add non-blocking communication content

### PR DESCRIPTION
**Pull request by mabraham**
For now, just seeing whether @robertodr's magic custom.py glossary works when building online. It doesn't do anything for me when I run sphinx-build manually locally.

### CODE REVIEW

**Comment by robertodr** [comment]
The GitHub action building the website does `make dirhtml`, whereas `sphinx-build` produces HTML. Internal references work differently between the `dirhtml` and `html` outputs: `custom.py` works with the former, since that is what will ultimately produce the website.

**Comment by mabraham** [comment]
> The GitHub action building the website does `make dirhtml`, whereas `sphinx-build` produces HTML. Internal references work differently between the `dirhtml` and `html` outputs: `custom.py` works with the former, since that is what will ultimately produce the website.

OK `make dirhtml` got the quick-reference building as I hoped, but my local `sphinx-build` it calls still reports undefined references for `|term-MPI_xxx|`. I'll see if the web does it right... Edit: Not yet :-(

**Comment by robertodr** [comment]
The logs only show two warnings: https://github.com/ENCCS/intermediate-mpi/actions/runs/400800349 Maybe they are not defined in `custom.py`?

**Comment by mabraham** [comment]
> The logs only show two warnings: https://github.com/ENCCS/intermediate-mpi/actions/runs/400800349 Maybe they are not defined in `custom.py`?

They are there, but I was trying to be fancy and make links to the MPI standard also. I'll try a simpler approach

**Comment by mabraham** [comment]
> > The logs only show two warnings: https://github.com/ENCCS/intermediate-mpi/actions/runs/400800349 Maybe they are not defined in `custom.py`?
> 
> They are there, but I was trying to be fancy and make links to the MPI standard also. I'll try a simpler approach

Should be warning-free now

**Comment by robertodr** [comment]
There are conflicts, I don't understand why though... Can you resolve them?

**Comment by mabraham** [comment]
> There are conflicts, I don't understand why though... Can you resolve them?

Yes, my patch in master did a proper job of changing custom.py, which is now integrated here also

File: content/non-blocking-communication-pt1.rst [hunk 1]
```diff
@@ -1,33 +1,307 @@
...
+                  int dest, int tag, MPI_Comm comm, MPI_Request *request)
+
+.. note::
+
+    ``buf``, ``count`` and ``datatype`` describe the buffer to be sent
+    to ``dest`` rank of ``comm`` using tag ``tag``. The ``request`` object
+    that is returned must be used to wait on the communication later.
+
+Other calls exist for other sending modes, including buffered,
+synchronous, and ready-mode sends.
```
Comment (robertodr) [inline content/non-blocking-communication-pt1.rst]: Do you think it would be appropriate to revive the table summarizing the various p2p functions?

Comment (mabraham) [reply content/non-blocking-communication-pt1.rst]: Yes if it's not anywhere else, it's a good reminder

File: content/non-blocking-communication-pt1.rst [hunk 2]
```diff
@@ -1,33 +1,307 @@
...
+
+
+Non-blocking MPI receive call
+-----------------------------
 
+An |term-MPI_Irecv| creates a receive request and returns a request
```
Comment (robertodr) [inline content/non-blocking-communication-pt1.rst]: How about: "An |term-MPI_Irecv| creates and returns a receive request in a `MPI_Request` object." To my non-native ear the current version sounds like the returned request is the return value.

File: content/non-blocking-communication-pt1.rst [hunk 3]
```diff
@@ -1,33 +1,307 @@
...
+
+An |term-MPI_Irecv| can be used to match any kind of send, regardless of
+sending mode or blocking status.
+
+Waiting for non-blocking call completion
+----------------------------------------
+
+An |term-MPI_Wait| call waits for completion of the operation that
+created the request object passed to it. For a send, the semantics of
+the sending mode have been restored (not necessarily that the message
```
Comment (robertodr) [inline content/non-blocking-communication-pt1.rst]: I don't understand what "the semantics of the sending mode have been restored" :disappointed:

Comment (mabraham) [reply content/non-blocking-communication-pt1.rst]: You're right, I meant more like "fulfilled"

File: content/non-blocking-communication-pt1.rst (hunk 2 above)
Comment (mabraham) [reply content/non-blocking-communication-pt1.rst]: Agreed

File: content/non-blocking-communication-pt1.rst [hunk 4]
```diff
@@ -1,33 +1,307 @@
...
+|term-MPI_Waitsome|, and |term-MPI_Waitall| for these use cases. For example,
+waiting for any request to complete may allow the caller to continue
+with related computation while waiting for other requests to complete.
+
+
+Testing for non-blocking call completion
+----------------------------------------
+
+An |term-MPI_Test| call returns immediately whether a corresponding
+|term-MPI_Wait| would return immediately. completes the operation that
```
Comment (robertodr) [inline content/non-blocking-communication-pt1.rst]: The sentence starting with "completes" might be missing a subject.

File: content/non-blocking-communication-pt1.rst [hunk 5]
```diff
@@ -1,33 +1,307 @@
...
+created the request object passed to it. For a send, the semantics of
+the sending mode have been restored (not necessarily that the message
+has been received). For a receive, the buffer is now valid for use,
+however the send has not necessarily completed (though obviously has
+been initiated).
```
Comment (robertodr) [inline content/non-blocking-communication-pt1.rst]: Is the copy-paste intended here?

File: content/non-blocking-communication-pt1.rst [hunk 6]
```diff
@@ -1,33 +1,307 @@
...
+    ``request`` describes the operation to be waited upon. ``status``
+    returns the status of that operation. If the status is not needed,
+    pass ``MPI_STATUS_IGNORE``.
+
+It can be efficient to wait on any one, some, or all of a set of
+operations before returning. MPI provides |term-MPI_Waitany|,
+|term-MPI_Waitsome|, and |term-MPI_Waitall| for these use cases. For example,
+waiting for any request to complete may allow the caller to continue
+with related computation while waiting for other requests to complete.
+
```
Comment (robertodr) [inline content/non-blocking-communication-pt1.rst]: I think this copy-paste was *not* intended

File: content/non-blocking-communication-pt1.rst [hunk 7]
```diff
@@ -1,33 +1,307 @@
...
+     
 See also
 --------
 
 
-* The lecture covering MPI+OpenMP from EPCC is available `here <http://www.archer.ac.uk/training/course-material/2020/01/advMPI-imperial/Slides/L06-MPIandOpenMP.pdf>`_
+* TODO
+* TODO
 * Chapter 6 of the **Using Advanced MPI** by William Gropp *et al.* :cite:`Gropp2014-dz`
```
Comment (robertodr) [inline content/non-blocking-communication-pt1.rst]: The "See also" is also not pertinent to this episode.

File: content/non-blocking-communication-pt1.rst (hunk 1 above)
Comment (robertodr) [reply content/non-blocking-communication-pt1.rst]: Here's the rst source:
```rst

.. table:: Point-to-point communication functions
   :widths: auto

   +------------------------+------------------------+------------------+-------------------+-----------------------+
   |                                                 |                    Communication                             |
   +                                                 +------------------+-------------------+-----------------------+
   |                                                 | Blocking         | Nonblocking       | Persistent            |
   +========================+========================+==================+===================+=======================+
   |                        | **Standard**           | |term-MPI_Send|  | |term-MPI_Isend|  | |term-MPI_Send_init|  |
   +                        +------------------------+------------------+-------------------+-----------------------+
   | **Mode**               | **Synchronous**        | |term-MPI_Ssend| | |term-MPI_Issend| | |term-MPI_Ssend_init| |
   +                        +------------------------+------------------+-------------------+-----------------------+
   |                        | **Ready**              | |term-MPI_Rsend| | |term-MPI_Irsend| | |term-MPI_Rsend_init| |
   +                        +------------------------+------------------+-------------------+-----------------------+
   |                        | **Buffered**           | |term-MPI_Bsend| | |term-MPI_Ibsend| | |term-MPI_Bsend_init| |
   +------------------------+------------------------+------------------+-------------------+-----------------------+
```

File: content/non-blocking-communication-pt1.rst (hunk 4 above)
Comment (mabraham) [reply content/non-blocking-communication-pt1.rst]: done

File: content/non-blocking-communication-pt1.rst (hunk 5 above)
Comment (mabraham) [reply content/non-blocking-communication-pt1.rst]: Good catch, I didn't clean up the copy-paste enough

File: content/non-blocking-communication-pt1.rst (hunk 6 above)
Comment (mabraham) [reply content/non-blocking-communication-pt1.rst]: Yep

File: content/non-blocking-communication-pt1.rst (hunk 7 above)
Comment (mabraham) [reply content/non-blocking-communication-pt1.rst]: 0k

**Comment by mabraham** [comment]
Thanks for the review. I've fixed the issues noted and added almost all the remaining content

File: content/non-blocking-communication-pt1.rst (hunk 1 above)
Comment (mabraham) [reply content/non-blocking-communication-pt1.rst]: I still need to work this in

Comment (mabraham) [reply content/non-blocking-communication-pt1.rst]: Worked this in, although left some TODO content in the quick-refernce, as I have to go make dinner!

**Comment by mabraham** [comment]
I'm happy to integrate this if others are!

File: content/non-blocking-communication-pt1.rst [hunk 8]
```diff
@@ -1,33 +1,324 @@
...
+   :widths: auto
+
+   +------------------------+------------------------+------------------+-------------------+
+   |                                                 |             Communication            |
+   +                                                 +------------------+-------------------+
+   |                                                 | Blocking         | Nonblocking       |
+   +========================+========================+==================+===================+
+   |                        | **Standard**           | |term-MPI_Send|  | |term-MPI_Isend|  |
+   +                        +------------------------+------------------+-------------------+
+   | **Mode**               | **Synchronous**        | |term-MPI_Ssend| | |term-MPI_Issend| |
```
Comment (robertodr) [inline content/non-blocking-communication-pt1.rst:155]: |term-MPI_Issend|, |term-MPI_Rsend|, |term-MPI_Irsend|, |term-MPI_Bsend|, and |term-MPI_Ibsend| are not defined. I think these can be simply monospace, no need to have them in the glossary.

**Review by robertodr — APPROVED** [review]

File: content/non-blocking-communication-pt1.rst (hunk 8 above)
Comment (mabraham) [reply content/non-blocking-communication-pt1.rst:155]: Also reasonable

**Pull request merged**